### PR TITLE
Prototype pollution fix - checking magic attributes

### PR DESCRIPTION
**Pull request by d3v53c**
### 📊 Metadata *

easydot is vulnerable to Prototype Pollution.

### ⚙️ Description *

Prototype Pollution refers to the ability to inject properties into existing JavaScript language construct prototypes, such as objects.
JavaScript allows all Object attributes to be altered, including their magical attributes such as proto, constructor and prototype.
An attacker manipulates these attributes to overwrite, or pollute, a JavaScript application object prototype of the base object by injecting other values.
Properties on the Object.prototype are then inherited by all the JavaScript objects through the prototype chain.

### 💻 Technical Description *

Fixed by avoiding setting magical attributes. The bug is fixed by validating the input strArray to check for prototypes. It is implemented by a simple validation to check for prototype keywords (proto, constructor and prototype), where if it exists, the function returns the object without modifying it, thus fixing the Prototype Pollution Vulnerability.

### 🐛 Proof of Concept (PoC) *

Create the following PoC file:

```
// poc.js
var easydot = require("easydot")
var obj = {}
const easy = easydot(obj);
console.log("Before : " + {}.polluted);
easy['__proto__.polluted']= 'Yes! Its Polluted';
console.log("After : " + {}.polluted)
```

Execute the following commands in terminal:

```
npm i easydot # Install affected module
node poc.js #  Run the PoC
```

Check the Output:

```
Before : undefined
After : Yes! Its Polluted
```

### 🔥 Proof of Fix (PoF) *

Before:
image

After:
image

### 👍 User Acceptance Testing (UAT)

image

After applying the fix, functionality is unaffected.

### 🔗 Relates to...

https://www.huntr.dev/bounties/1-npm-easydot/

### CODE REVIEW

**Review by mzfr — COMMENTED** [review]
The fix is good but I think [they](https://github.com/418sec/easydot/pull/1) made the PR first.

**Comment by d3v53c** [comment]
> The fix is good but I think [they](https://github.com/418sec/easydot/pull/1) made the PR first.

@mzfr , I'm aware of that. I've updated syntax issues in testcases as well, and probably a cleaner fix added, I think.

Cheers!!

File: src/index.js [hunk 1]
```diff
@@ -44,6 +54,7 @@ export default (target, safeOverJump = false) => {
     const lastKey = keys.pop();
     let tmp = target;
     keys.forEach(key => {
+      if (isPrototypePolluted(key)) return;
```
Comment (Mik317) [inline src/index.js:57]: Do you think is better returning a `false` object instead of void?

**Review by Mik317 — APPROVED** [review]
LGTM :smile: :cake:
Preferring this approach actually, maybe `continue` in the other PR isn't the best fit for the `forEach`

Cheers,
Mik

**Comment by huntr-helper** [comment]
Sorry d3v53c, we enjoyed reviewing your fix but it has not been selected this time. If this bounty has not been closed, please feel free to try again with a new pull request!

We appreciate your effort and look forward to reviewing more of your fixes in the future! 🔨😎

[Come join us on Discord](https://huntr.dev/discord)

File: src/index.js (hunk 1 above)
Comment (d3v53c) [reply src/index.js:57]: `false` would've been a better option.

**Comment by d3v53c** [comment]
Updated according to comments.

**Pull request closed without merging**